### PR TITLE
Leopoly_Phase1_004_RayPick-Changes

### PR DESCRIPTION
**Pull request by humbletim**
Changes to the Ray Intersection to support more detailed information coming back from intersections to make it easier to work with mesh data.  Example detail includes submesh index, submesh name, and effective triangle face. 

No known impact on existing application behavior or ray picking.  A few places in the old `script-archive` made reference to `.extraInfo` and have been updated to reflect the new structure.

#### Testing

* General smoke test (exercising mouse and hand laser picking in general).
(Make sure to cover every kind of Entity type for results)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5776/app_12037_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5776/HighFidelity-Beta-PR12037-a146b778f784ce6f06ee86fdafd4f36c4c05fea5.dmg

**Comment by MiladNazeri** [comment]
Overall Leopoly PR roadmap:
https://docs.google.com/document/d/1GJC6qYHICYccIqdmaGgtt30Nj12Q3FtFmejns7biyks/edit
This is 4/9

Labeling Do not merge until code is reviewed for #'s 1-8

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 1]
```diff
@@ -756,7 +759,7 @@ void EntityTreeRenderer::mouseReleaseEvent(QMouseEvent* event) {
     // Even if we're no longer intersecting with an entity, if we started clicking on it, and now
     // we're releasing the button, then this is considered a clickReleaseOn event
     if (!_currentClickingOnEntityID.isInvalidID()) {
-        glm::vec2 pos2D = projectOntoEntityXYPlane(rayPickResult.entity, ray, rayPickResult);
+        glm::vec2 pos2D = projectOntoEntityXYPlane( entity, ray, rayPickResult);
```
Comment (SamGondelman) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: extra space

File: libraries/entities/src/EntityTreeElement.cpp [hunk 2]
```diff
@@ -588,57 +588,60 @@ bool EntityTreeElement::bestFitBounds(const glm::vec3& minPoint, const glm::vec3
...
+EntityItemID EntityTreeElement::findDetailedRayIntersection(const glm::vec3& origin, const glm::vec3& direction, bool& keepSearching,
                                     OctreeElementPointer& element, float& distance, BoxFace& face, glm::vec3& surfaceNormal,
                                     const QVector<EntityItemID>& entityIdsToInclude, const QVector<EntityItemID>& entityIDsToDiscard,
-                                    bool visibleOnly, bool collidableOnly, void** intersectedObject, bool precisionPicking, float distanceToElementCube) {
+                                    bool visibleOnly, bool collidableOnly, QVariantMap& extraInfo, bool precisionPicking, float distanceToElementCube) {
 
     // only called if we do intersect our bounding cube, but find if we actually intersect with entities...
     int entityNumber = 0;
-    bool somethingIntersected = false;
+    EntityItemID somethingIntersected;
```
Comment (SamGondelman) [inline libraries/entities/src/EntityTreeElement.cpp]: Call this entityID instead of somethingIntersected

File: libraries/fbx/src/FBXReader.cpp [hunk 3]
```diff
@@ -1827,7 +1830,19 @@ FBXGeometry* FBXReader::extractFBXGeometry(const QVariantHash& mapping, const QS
             }
         }
     }
-
+    {
+        int i=0;
```
Comment (SamGondelman) [inline libraries/fbx/src/FBXReader.cpp]: spaces around `=`

File: libraries/render-utils/src/Model.cpp [hunk 4]
```diff
@@ -397,18 +401,36 @@ bool Model::findRayIntersectionAgainstSubMeshes(const glm::vec3& origin, const g
                     bestDistance = worldDistance;
                     intersectedSomething = true;
                     face = triangleSetFace;
-                    surfaceNormal = glm::vec3(meshToWorldMatrix * glm::vec4(triangleSetNormal, 0.0f));
-                    extraInfo = geometry.getModelNameOfMesh(subMeshIndex);
+                    bestModelTriangle = triangleSetTriangle;
+                    bestWorldTriangle = triangleSetTriangle * meshToWorldMatrix;
+                    extraInfo["_worldIntersectionPoint"] = vec3toVariant(worldIntersectionPoint);
+                    extraInfo["_meshIntersectionPoint"] = vec3toVariant(meshIntersectionPoint);
```
Comment (SamGondelman) [inline libraries/render-utils/src/Model.cpp]: Should these have underscores?  Also, make sure to document the new data in extraInfo.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5921/app_12037_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5922/app_12037_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5922/HighFidelity-Beta-PR12037-4021455058486c7be9c0fd7cf8d63348e0d522ec.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5922/HighFidelity-Beta-PR12037-4021455058486c7be9c0fd7cf8d63348e0d522ec.dmg

**Comment by ZappoMan** [comment]
This PR appears to change an API... namely the old "extraInfo" used to be the sub-mesh name from the model... now it's possible that's been broken for a while, but we should make sure that any samples and documentation is updated.... AND... we really should follow our backward compatibility process.

@misslivirose @tapalisa @samcake @divyaelango88

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6060/app_12037_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6060/HighFidelity-Beta-PR12037-990f1db831fdf31f1aeef61d98e6a550fb88887b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6060/HighFidelity-Beta-PR12037-990f1db831fdf31f1aeef61d98e6a550fb88887b.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6544/app_12037_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6545/app_12037_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6545/HighFidelity-Beta-PR12037-9fb20bee805beb8016a8a0248283c4957cd97bf8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6545/HighFidelity-Beta-PR12037-9fb20bee805beb8016a8a0248283c4957cd97bf8.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6652/app_12037_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6652/HighFidelity-Beta-PR12037-e2afcac53427a5fceeefc96ecfc25146fef65dae.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6652/HighFidelity-Beta-PR12037-e2afcac53427a5fceeefc96ecfc25146fef65dae.exe

**Comment by sbennett77** [comment]
Currently Testing.

**Comment by sbennett77** [comment]
Passed.

@conklin94122

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6679/app_12037_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6679/HighFidelity-Beta-PR12037-e2afcac53427a5fceeefc96ecfc25146fef65dae.dmg

**Comment by MiladNazeri** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6686/app_12037_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6686/HighFidelity-Beta-PR12037-e2afcac53427a5fceeefc96ecfc25146fef65dae.dmg

**Comment by MiladNazeri** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6691/app_12037_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6691/HighFidelity-Beta-PR12037-e2afcac53427a5fceeefc96ecfc25146fef65dae.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6691/HighFidelity-Beta-PR12037-e2afcac53427a5fceeefc96ecfc25146fef65dae.exe

**Pull request merged**